Refactor Graphene out of CNode

ptschip opened the pull request.

ptschip committed:
Add CGrpaheneBlock pointer to CBlockThinRelay

ptschip committed:
Move nSizeGrapheneBlock out of CNode

ptschip committed:
Move grapheneBlockWaitingForTxns out of CNode

ptschip committed:
Move nLocalGrapheneBlockBytes out of CNode

ptschip committed:
Move grapheneBlockHashes out of CNode

ptschip committed:
Move mapGrapheneMissingTx out of CNode

ptschip committed:
Move mapGrapheneAdditionalTxs out of CNode

ptschip committed:
Move mapGrapheneHashOrderIndex out of CNode

ptschip committed:
Move grapheneBlock out of CNode

ptschip committed:
Make pointer all shared_ptrs

ptschip committed:
Do not use references to smart pointers

ptschip committed:
Fix bug in RequestFailOverBlock()

ptschip committed:
Reject blocks by current size only during reconstruction

Since we only used shared pointers for transactions in the mempool, and only
begin recontruction of the block once we have all the necessary data,
it is no longer needed to be able to track bytes used for
multiple thintype blocks that are still in flight and only partially
reconstructed. This commit simplifies the entire procedure and simply
checks if the block has gone over the byte limit during reconstuction and
if so then cleanup the data and disconnet the peer.

sickpig commented:
@ptschip I saw you disabled `excessive` an `mempool_accept` tests, is this because PR is in WIP? Should be wait before reviewing?

ptschip commented:
oops, that was unintentional...i'll re-enable them

On 29/04/2019 5:49 a.m., Andrea Suisani wrote:
>
> @ptschip <https://github.com/ptschip> I saw you disabled |excessive|
> an |mempool_accept| tests, is this because PR is in WIP? Should be
> wait before reviewing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1708#issuecomment-487566522>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUBK76CXKKKAHDOCEN3PS3VGPANCNFSM4HI6X2JA>.
>

ptschip committed:
Re-enable excessive.py and mempool_accept.py

gandrewstone commented on src/blockrelay/graphene.cpp:
```diff
@@ -109,12 +105,15 @@ bool CGrapheneBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
     vRecv >> grapheneBlockTx;
 
+    auto pblock = thinrelay.GetBlockToReconstruct(pfrom);
+    std::shared_ptr<CGrapheneBlock> grapheneBlock = pblock->grapheneblock;
```
make sure pblock->grapheneblock != nullptr since SetNull/ClearAllBlockData can set it to null.

ptschip replied:
adding a DbgAssert here.

gandrewstone commented on src/blockrelay/graphene.cpp:
```diff
@@ -337,38 +336,39 @@ bool CGrapheneBlock::CheckBlockHeader(const CBlockHeader &block, CValidationStat
 
-    CGrapheneBlock grapheneBlock(NegotiateGrapheneVersion(pfrom), NegotiateFastFilterSupport(pfrom));
-    vRecv >> grapheneBlock;
+    std::shared_ptr<CGrapheneBlock> grapheneBlock = pblock->grapheneblock;
```
This might be picky, but if you do:
globalVar = malloc()
localVar = globalVar
you risk another thread modifying globalVar in between those 2 lines.  Maybe that can't happen here anyway due to locks, BUT if you do:
localVar = malloc()
globalVar = localVar

the code shows that it can't happen without knowing anything about the rest of program state.   So I think doing it that way is better.

ptschip replied:
I modified SetBlockToReconstruct() so that once it's set for a particular node and hash it won't be re-intitialized by any susequent attempts.   At any rate , the way it was we were holding the smart pointer already to graphenblock so if another thread modified it in pblock we would still have our own separate smart point to our data so it wouldn't impact the processing of HandleMessage...but if there was a re-request, then you'd end up using the new pointer to grapheneblock and potentially messing up your reconstruction, but it wouldn't crash...but for now it's all moot because we would never download two grapheneblocks for the same blockhash anyway, and furthermore we can't currently download two grapheneblocks at the same time...Regardless you bring up good points that need to be considered and I think the changes in SetBlockToReconstruct() take care of our current needs.

gandrewstone commented on src/blockrelay/graphene.cpp:
```diff
@@ -431,42 +431,31 @@ bool CGrapheneBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, std::string
-    graphenedata.ClearGrapheneBlockData(pfrom);
-    pfrom->nSizeGrapheneBlock = nSizeGrapheneBlock;
+    std::shared_ptr<CGrapheneBlock> grapheneBlock = pblock->grapheneblock;
 
```
validate grapheneBlock != nullptr

gandrewstone replied:
also is this == pblock->grapheneblock?

I think this routine sometimes uses this and sometimes grapheneBlocks and may assume they are the same.  One idea is to DbgAssert(this == pfrom->grapheneBlock, return false)

ptschip replied:
I'll take out graphenBlock, it's a bit of a circular reference...we only need to use "this"...that will also get rid of need to check for nullptr in this particular case.

ptschip replied:
added checks for != nullptr and took out the references to "this" and just standardized across all thinrelay types in the same fashion.

ptschip committed:
Use reset() instead of setting to nullptr in SetNull()

This prevents us from ever having other threads possibly setting
our thintype objects to  nullptr in CBlockThinRelay objects. If some
other thread is still holds a shared pointer then is won't be
invalidated and the destructor won't be called.

ptschip committed:
When setting a block to reconstruct check first if we already have one

It's possible that we could have both a thinblock and grapheneblock
from being requested for the same block hash so check first if we've
already create a shared pointer to a CBlockThinRelay object and only
if we don't have one then create a fresh instance. This way we won't
accidentally remove an instance which has an active thintype block
in use.

ptschip commented:
@gandrewstone updated as per your comments and rebased.

ptschip committed:
Add a few DbgAsserts

Also check if ptx is null before updating block bytes

 In practice getting a null ptx could not happen because we always
 make sure we have all the transactions before we reconstruct the block
 however just in case someone makes a mistake in the future it's a good
 check to have.

ptschip committed:
Standardize handling of smart pointer references in all thinrelay types

ptschip committed:
Combine both thinblock and xthinblock mapMissingTx

Since we no longer use nullptr for xthinblock or thinblock which are held
in pblock then we have to either pass in the netmessage type or simply
combine both maps into one. Here, the simple approach of combining both
maps is done which prevents any issues regarding passing in an incorrect
netmessage type.
